inst/doc/pitfalls.Rmd

```diff
@@ -0,0 +1,51 @@
+---
+title: "Common Pitfalls and Solutions"
+author: "Alison Appling"
+date: "February 4, 2019"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Shared cache}
+  %\VignetteEngine{knitr::rmarkdown}
+  \usepackage[utf8]{inputenc}
+---
+
+<head>
+<style>
+sym {color: red;}
+sol {color: blue;}
+</style>
+</head>
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, cache=FALSE, collapse=TRUE)
+```
+
+The following bullets describe problems or frustrations that `scipiper` users might encounter, along with some ideas for how to interpret and fix those issues. Each "pitfall" is presented in this format: _<sym>Problem or symptom.</sym> Explanation. <sol>Solution.</sol>_
+
+## General use of scipiper/remake
+
+* <sym>Unnecessary re-runs of functions.</sym> `scipiper` and `remake` don't update their databases when you run a function outside of scipiper, even if the function creates the desired outputs. This may be frustrating if you've just run a long-running function outside of `scipiper` and are now attempting to use `scipiper` to build targets that depend on the function's output, because `scipiper` will insist on running that function again. <sol>Anytime you might want the output of a function to be considered complete, build the `scipiper` target rather than calling the function directly.</sol>
+
+* <sym>Common debugging methods don't work.</sym> If you build a `scipiper` target that runs function `A()` to create target `a`, setting RStudio checkpoints or `debug(A)` won't let you debug function `A` when you call `scmake('a')`. <sol>Temporarily insert a call to `browser()` within the function instead.</sol>
+
+* <sym>The GitHub repo is too big.</sym> You might be committing data files or large task remake files to the repo. <sol>In general, you can safely .gitignore all data files, the .remake/ folder, and (optionally) any task remake files.</sol>
```

Comment:
Creating `tmp/` directories for intermediate data processing steps and .gitignoring those folders is also common

Reply:
You can only gitignore data files if you are using a shared cache approach, right? Or you expect it to build fairly quickly?

Reply:
No, I think you can and should generally gitignore data files for all projects, because GitHub (without GitLFS, anyway) isn't made to store big or binary files, and most data files are big and binary.

But if you don't use a shared cache, it's on you to store and back up the data files in some other way - could be as simple/risky as keeping the canonical copy in your local repo, or could involve doing the whole analysis on a shared drive, or running backups yourself, or manually copying files to and from some shared location. So I'll edit this text to add that caveat to the "safely" idea - data do need to be backed up somewhere, it's just that GitHub is usually a poor choice.

The one situation where I think it's a good idea to NOT .gitignore a data file is when it's (1) small, (2) ideally text-based so that the git diffs are meaningful to humans, and (3) possibly also not practical to store anywhere better. So we often make this exception for configuration files, small input data files that would be a bother to share via S3 or Drive, small input files when we're not doing any sharing elsewhere, and occasionally some files that summarize the status of the pipeline or the results as those change over time. We occasionally also make the exception for image files that we really want to see and store multiple versions of (e.g., you probably did this with gsplot?) - it's not that you _can't_ store binary files but that their storage often makes the repo too darn big.